src/main/java/net/minestom/server/listener/preplay/HandshakeListener.java

```diff
@@ -38,93 +38,99 @@ public final class HandshakeListener {
     private static final Component INVALID_BUNGEE_FORWARDING = Component.text("Invalid connection, please connect through the BungeeCord proxy. If you believe this is an error, contact a server administrator.", NamedTextColor.RED);
 
     public static void listener(@NotNull ClientHandshakePacket packet, @NotNull PlayerConnection connection) {
-        String address = packet.serverAddress();
         switch (packet.intent()) {
             case STATUS -> connection.setConnectionState(ConnectionState.STATUS);
-            case LOGIN -> {
+            case LOGIN, TRANSFER -> {
                 connection.setConnectionState(ConnectionState.LOGIN);
-                if (packet.protocolVersion() != MinecraftServer.PROTOCOL_VERSION) {
-                    // Incorrect client version
-                    disconnect(connection, INVALID_VERSION_TEXT);
-                }
+                handlePlayerLogin(packet, connection);
+            }
+        }
+
+        if (connection instanceof PlayerSocketConnection) {
+            // Give to the connection the server info that the client used
+            ((PlayerSocketConnection) connection).refreshServerInformation(
+                    packet.serverAddress(),
+                    packet.serverPort(),
+                    packet.protocolVersion());
+        }
```

Comment:
You can invert the condition to reduce nesting here.
```suggestion
        if (!(connection instanceof PlayerSocketConnection connection)) return;
        // Give to the connection the server info that the client used
        connection.refreshServerInformation(
                packet.serverAddress(),
                packet.serverPort(),
                packet.protocolVersion()
        );
```